File: core/src/main/java/com/vzome/core/exporters/DxfExporter.java
package com.vzome.core.exporters;

import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.Writer;
import java.text.NumberFormat;
import java.util.Locale;

import com.vzome.core.algebra.AlgebraicVector;
import com.vzome.core.math.RealVector;
import com.vzome.core.model.Manifestation;
import com.vzome.core.model.Strut;
import com.vzome.core.render.RenderedManifestation;


public class DxfExporter extends GeometryExporter
{
    @Override
	public void doExport( File directory, Writer writer, int height, int width ) throws IOException
	{
        output = new PrintWriter( writer );
        output .println( "0" );
        output .println( "SECTION" );
        output .println( "2" );
        output .println( "ENTITIES" );
		
        NumberFormat format = NumberFormat .getNumberInstance( Locale .US );
        format .setMaximumFractionDigits( 6 );
        double inchScaling = mModel .getCmScaling() / 2.54d;

        for (RenderedManifestation rm : mModel) {
            Manifestation man = rm .getManifestation();
            if ( man instanceof Strut ) {
                output .println( "0" );
                output .println( "LINE" );
                output .println( "8" );
                output .println( "vZome" ); // this is the "layer" the line appears in; it need not be predefined
                AlgebraicVector start = ((Strut) man) .getLocation();
                AlgebraicVector end = ((Strut) man) .getEnd();
                RealVector rv = mModel .renderVector( start );
                rv = rv .scale( inchScaling );
                output .println( "10" );
                output .println( format.format( rv .x ) );
                output .println( "20" );
                output .println( format.format( rv .y ) );
                output .println( "30" );
                output .println( format.format( rv .z ) );
                rv = mModel .renderVector( end );
                rv = rv .scale( inchScaling );
                output .println( "11" );
                output .println( format.format( rv .x ) );
                output .println( "21" );
                output .println( format.format( rv .y ) );
                output .println( "31" );
                output .println( format.format( rv .z ) );
            }
        }
        
        output .println( "0" );
        output .println( "ENDSEC" );
        output .println( "0" );
        output .println( "EOF" );

		output .flush();
	}


    @Override
    public String getFileExtension()
    {
        return "dxf";
    }

}

File: online/src/worker/legacy/from-java/com/vzome/core/exporters/DxfExporter.ts
import { java, javaemul } from "../../../../../candies/j4ts-2.1.0-SNAPSHOT/bundle.js";
import { AlgebraicVector } from "../algebra/AlgebraicVector.js";
import { GeometryExporter } from "./GeometryExporter.js";
import { RealVector } from "../math/RealVector.js";
import { Manifestation } from "../model/Manifestation.js";
import { Strut } from "../model/Strut.js";
import { File } from "../../../../java/io/File.js";
import { PrintWriter } from "../../../../java/io/PrintWriter.js";
import { NumberFormat } from "../../../../java/text/NumberFormat.js";

export class DxfExporter extends GeometryExporter {
    /**
     * 
     * @param {File} directory
     * @param {java.io.Writer} writer
     * @param {number} height
     * @param {number} width
     */
    public doExport(directory: File, writer: java.io.Writer, height: number, width: number) {
        this.output = new PrintWriter(writer);
        this.output.println$java_lang_Object("0");
        this.output.println$java_lang_Object("SECTION");
        this.output.println$java_lang_Object("2");
        this.output.println$java_lang_Object("ENTITIES");
        const format: NumberFormat = NumberFormat.getNumberInstance(java.util.Locale.US);
        format.setMaximumFractionDigits(6);
        const inchScaling: number = this.mModel.getCmScaling() / 2.54;
        for(let index=this.mModel.iterator();index.hasNext();) {
            let rm = index.next();
            {
                const man: Manifestation = rm.getManifestation();
                if (man != null && (man.constructor != null && man.constructor["__interfaces"] != null && man.constructor["__interfaces"].indexOf("com.vzome.core.model.Strut") >= 0)){
                    this.output.println$java_lang_Object("0");
                    this.output.println$java_lang_Object("LINE");
                    this.output.println$java_lang_Object("8");
                    this.output.println$java_lang_Object("vZome");
                    const start: AlgebraicVector = (<Strut><any>man).getLocation();
                    const end: AlgebraicVector = (<Strut><any>man).getEnd();
                    let rv: RealVector = this.mModel.renderVector(start);
                    rv = rv.scale(inchScaling);
                    this.output.println$java_lang_Object("10");
                    this.output.println$java_lang_Object(format.format(rv.x));
                    this.output.println$java_lang_Object("20");
                    this.output.println$java_lang_Object(format.format(rv.y));
                    this.output.println$java_lang_Object("30");
                    this.output.println$java_lang_Object(format.format(rv.z));
                    rv = this.mModel.renderVector(end);
                    rv = rv.scale(inchScaling);
                    this.output.println$java_lang_Object("11");
                    this.output.println$java_lang_Object(format.format(rv.x));
                    this.output.println$java_lang_Object("21");
                    this.output.println$java_lang_Object(format.format(rv.y));
                    this.output.println$java_lang_Object("31");
                    this.output.println$java_lang_Object(format.format(rv.z));
                }
            }
        }
        this.output.println$java_lang_Object("0");
        this.output.println$java_lang_Object("ENDSEC");
        this.output.println$java_lang_Object("0");
        this.output.println$java_lang_Object("EOF");
        this.output.flush();
    }

    /**
     * 
     * @return {string}
     */
    public getFileExtension(): string {
        return "dxf";
    }
}
DxfExporter["__class"] = "com.vzome.core.exporters.DxfExporter";
